tools/c7n_azure/c7n_azure/filters.py

```diff
@@ -184,15 +187,19 @@ def get_metric_data(self, resource):
         cached_metric_data = self._get_cached_metric_data(resource)
         if cached_metric_data:
             return cached_metric_data['measurement']
-
-        metrics_data = self.client.metrics.list(
-            resource['id'],
-            timespan=self.timespan,
-            interval=self.interval,
-            metricnames=self.metric,
-            aggregation=self.aggregation,
-            filter=self.filter
-        )
+        try:
+            metrics_data = self.client.metrics.list(
```

Comment:
Can you put a breakpoint on this exception handler and get us the full exception object? It'll be some Azure error type.  I'd like to see what it is exactly so we can try to focus our error handling here.

Reply:
So whether the exception is because the resource has no metrics or just because the user spelt the metric wrong it is the same type: `azure.mgmt.monitor.v2018_01_01.models._models_py3.ErrorResponseException`

I originally did not check for this specific Exception and resorted to catching all exceptions because I didn't want to hardcode the solution to this specific version of the api.

The thing that does change between the two is the error string:
```
# Resource with no available metrics
"Failed to find metric configuration for provider: Microsoft.Web, resource Type: serverfarms, metric: CpuPercentagegjk, Valid metrics: "

# Resource with available metrics but wrong metric name
"Failed to find metric configuration for provider: Microsoft.Web, resource Type: serverfarms, metric: CpuPercentagegjk, Valid metrics: CpuPercentage,MemoryPercentage,DiskQueueLength,HttpQueueLength,BytesReceived,BytesSent,TcpSynSent,TcpSynReceived,TcpEstablished,TcpFinWait1,TcpFinWait2,TcpClosing,TcpCloseWait,TcpLastAck,TcpTimeWait"
```
So by parsing the Valid metrics portion of the error message we can tell which of the scenarios we are in. Basically:
```
except Exception as e:
    error_message = str(e)
    valid_metrics = error_message.split("Valid metrics :")
    available_metrics = valid_metrics.split(",")
    if len(available_metrics) > 1:
        # Case where there is a list of available metrics
        raise Exception(f"Invalid metric name: {user_input_metric}. For this resource please choose from {valid_metrics}")
    
    self.log.warning("could not get metric:%s on %s" % (
        self.metric, resource['id']))
return None
```

This way we can choose the response based on which of the two scenarios we are in

Reply:
I spoke with @logachev and we decided to filter the except clause to the specific exception (provided by the instance of the client, not hardcoded for the specific api version) and then change the warning log messages to error log messages. We will not implement the string parsing to differentiate between the two cases here

Reply:
Yeah the right thing to do here is to have both of these cases warn and continue on.  

If we leave it how it was then problematic resources in a subscription can make custodian 100% non-functional, whereas the alternative is harmlessly logging a copy of the error for each resource rather than failing fast.

Reply:
I just pushed edits:
```
except self.client.models(self.client._get_api_version('metrics')).ErrorResponseException:
            self.log.error("could not get metric:%s on %s" % (
                self.metric, resource['id']))
            return None
```

Reply:
Should we include the list of available metrics in the error log then?

Reply:
Yes, I think we can include entire error message from Azure

Reply:
Ok the full error message from Azure is there. I had to split getting the Exception into two lines due to the linter